test/tests.py

```diff
@@ -1508,6 +1509,154 @@ def test(self):
                              join(testdir, "1.4", "misc", "dead_methods.dml")))
 
 
+func_start_re = re.compile(r'^static +\S+ +(\S+)\(.*\)$')
+begin_linemarks_func_re = re.compile(r'^// BEGIN-LINEMARKS-FUNC((?:\s+\S+)+)$')
+
+class Linemarks(CTestCase):
+    __slots__ = ()
+
+    def dml_lines_iter(self, c_file, c_lines_iter):
+        reset = True
+        dml_lineno = None
+        curr_func = None
+        for (c_lineno, line) in enumerate(c_lines_iter, 1):
+            if dml_lineno is not None:
+                dml_lineno += 1
+            line = line.rstrip()
+            m = line_directive_re.match(line)
+            if m is not None:
+                (lineno, filename) = m.groups()
+                lineno = int(lineno)
+                if (Path(self.scratchdir) / filename == c_file):
+                    assert not reset, \
+                        f"Redundant line directive reset at {c_lineno}"
+                    assert lineno == c_lineno + 1, \
+                        f"Incorrect line directive reset at {c_lineno}"
+                    reset = True
+                    dml_lineno = None
+                elif filename.endswith(self.filename):
+                    assert dml_lineno != lineno, \
+                            (f'Redundant linemark for {dml_lineno} at '
+                             + f'{c_lineno}')
+                    reset = False
+                    dml_lineno = lineno - 1
+                else:
+                    reset = False
+                    dml_lineno = None
+                continue
+            m = func_start_re.match(line)
+            if m is not None:
+                assert curr_func is None
+                curr_func = m.group(1)
+                assert reset, \
+                    f'Line directive not reset by start of {curr_func}'
+                continue
+            if dml_lineno is not None:
+                yield (dml_lineno, c_lineno, curr_func)
+            if curr_func is not None and line == '}':
+                curr_func = None
+
+        assert reset, "Line directive not reset by end of C file"
+        assert curr_func is None, "Unterminated C function"
+
+
+    def test(self):
+        super().test()
+        c_file = Path(f'{self.cfilename}.c').resolve()
+        h_file = Path(f'{self.cfilename}.h').resolve()
+        base_file = Path(self.filename).resolve()
+
+        func_spans = {}
+        cblock_spans = {}
+        not_linemarked = set()
```

Comment:
Please add comment and/or type annotations.

Reply:
oo! I can hear Python gurus all across the world becoming _absolutely infuriated_ by just the gall of _demanding_ people declare their types ;)

I've mentioned previously how much I hate type annotations due to the half-measure they represent, as they are nothing more but glorified comments, and so have an instinctive, irrational repulsion to their use. But since the question here is the choice between comments or glorified comments, I suppose I could indulge in the latter.

Reply:
Hence 'and/or'. I also don't quite like how type annotations turned out, but type annotations are often more concise than a comment in expressing what you want. Expressing `func_span: dict[str, set[(int, int)]] = {}` in text is much more  verbose without really adding anything essential.